### DISCUSSION TITLE
Using rate limiting with NextJS api routes

### DISCUSSION THREAD

**Discussion by LGmatrix13** [vercel/next.js#23515] [Help] [answered]
So, I have this api route:

` ` `
import dbConnect from '../../../utils/dbConnect';
import Post from '../../../models/Post';

dbConnect();

export default async (req, res) => {
    const { method } = req;

    switch (method) {
        case 'GET':
            try {
                const posts = await Post.find({});

                res.status(200).json({ success: true, data: posts })
            } catch (error) {
                res.status(400).json({ success:
false });
            }
            break;
        default:
            res.status(400).json({ success: false });
            break;
    }
}
` ` `

Is it possible to use an express module like express rate limit to implement rate limiting?

Nextjs has a tutorial for adding express model ware, but I’m still a little confused...

https://nextjs.org/docs/api-routes/api-middlewares

**Comment by wwwdepot** [comment, marked as answer]
Try https://github.com/vercel/next.js/tree/canary/examples/api-routes-rate-limit

**Reply by PlopTheReal** [reply]
Regarding the rate limiting example, how do you use the same LRU cache accross multiple API endpoints?

**Reply by LGmatrix13** [reply]
Maybe this is what you are looking for? #10729

https://github.com/vercel/next.js/discussions/10729

**Reply by LGmatrix13** [reply]
Maybe this is what you are looking for? #10729

https://github.com/vercel/next.js/discussions/10729

**Reply by PlopTheReal** [reply]
Thanks for the feedback but I don't think a cache-control header is going to stop bad actors from hammering the APIs.

**Reply by LGmatrix13** [reply]
Yeah, it is only applicable for `GET` requests. Maybe you can use the example by @wwwdepot, make it into some kind of middleware, and then append it to your various api endpoints. It seems that you can also use `express-rate-limit` package via middleware as well. See: https://nextjs.org/docs/api-routes/api-middlewares
